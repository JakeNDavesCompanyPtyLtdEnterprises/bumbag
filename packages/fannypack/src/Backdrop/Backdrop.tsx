import * as React from 'react';
import * as PropTypes from 'prop-types';
import { BackdropProps as ReakitBackdropProps } from 'reakit/ts/Backdrop/Backdrop';

import _Backdrop from './styled';
import {
  AnimateProps,
  animatePropTypes,
  animateDefaultProps,
  RestrictHideProps,
  restrictHidePropTypes,
  restrictDefaultProps
} from '../types';

export type LocalBackdropProps = {
  use?: any;
  children?: React.ReactNode;
  className?: string;
  /** Whether or not to show the component */
  isVisible?: boolean;
  onClick?(): void;
} & AnimateProps &
  RestrictHideProps;
export type BackdropProps = ReakitBackdropProps & LocalBackdropProps;

export const Backdrop: React.FunctionComponent<LocalBackdropProps> = ({ children, isVisible, ...props }) => (
  <_Backdrop visible={isVisible} {...props}>
    {children}
  </_Backdrop>
);

export const backdropPropTypes = {
  use: PropTypes.any,
  children: PropTypes.node,
  className: PropTypes.string,
  isVisible: PropTypes.bool,
  onClick: PropTypes.func,
  ...animatePropTypes,
  ...restrictHidePropTypes
};
Backdrop.propTypes = backdropPropTypes;

export const backdropDefaultProps = {
  use: undefined,
  children: undefined,
  className: undefined,
  isVisible: false,
  onClick: undefined,
  ...animateDefaultProps,
  ...restrictDefaultProps
};
Backdrop.defaultProps = backdropDefaultProps;

const C: React.FunctionComponent<BackdropProps> = Backdrop;
export default C;
