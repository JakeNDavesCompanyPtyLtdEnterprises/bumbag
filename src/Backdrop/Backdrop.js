// @flow
import React, { type Node } from 'react';

import _Backdrop from './styled';

type Props = {
  animated?: boolean,
  use?: any,
  children: Node,
  className?: string,
  /** Delay of the animation if one is specified. */
  delay?: string,
  /** Duration of the animation if one is specified. */
  duration?: string,
  /** Will the backdrop have an expand animation when it is toggled on/off? */
  expand?: boolean | string,
  /** Will the backdrop have a fade animation when it is toggled on/off? */
  fade?: boolean,
  /** Should the backdrop be hidden when 'esc' is pressed?  */
  hideOnEsc?: boolean,
  /** Should the backdrop be hidden when outside is clicked?  */
  hideOnClickOutside?: boolean,
  /** Whether or not to show the backdrop component */
  isVisible?: boolean,
  /** Will the backdrop have a slide animation when it is toggled on/off? */
  slide?: boolean | string,
  /** Timing of the animation */
  timing?: string
};

const Backdrop = ({ children, isVisible, ...props }: Props) => (
  <_Backdrop visible={isVisible} {...props}>
    {children}
  </_Backdrop>
);

Backdrop.defaultProps = {
  animated: false,
  use: undefined,
  className: undefined,
  delay: undefined,
  duration: '250ms',
  expand: undefined,
  fade: false,
  hideOnEsc: true,
  hideOnClickOutside: true,
  isVisible: false,
  slide: false,
  timing: 'ease-in-out'
};

export default Backdrop;
