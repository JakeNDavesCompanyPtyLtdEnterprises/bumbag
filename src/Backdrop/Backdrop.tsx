import * as React from 'react';
// @ts-ignore
import PropTypes from 'prop-types';

import _Backdrop from './styled';
import {
  AnimateProps,
  animatePropTypes,
  animateDefaultProps,
  RestrictHideProps,
  restrictHidePropTypes,
  restrictDefaultProps
} from '../types';

export interface BackdropProps {
  use?: any;
  children?: React.ReactNode;
  className?: string;
  /** Whether or not to show the component */
  isVisible?: boolean;
  onClick?(): void;
}

export const Backdrop: React.SFC<BackdropProps & AnimateProps & RestrictHideProps> = ({
  children,
  isVisible,
  ...props
}) => (
  <_Backdrop visible={isVisible} {...props}>
    {children}
  </_Backdrop>
);

Backdrop.propTypes = {
  use: PropTypes.any,
  children: PropTypes.node,
  className: PropTypes.string,
  isVisible: PropTypes.bool,
  onClick: PropTypes.func,
  ...animatePropTypes,
  ...restrictHidePropTypes
};
Backdrop.defaultProps = {
  use: undefined,
  children: undefined,
  className: undefined,
  isVisible: false,
  onClick: undefined,
  ...animateDefaultProps,
  ...restrictDefaultProps
};

export default Backdrop;
