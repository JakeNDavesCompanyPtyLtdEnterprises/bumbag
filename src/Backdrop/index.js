export { default } from './Backdrop';
export { default as Backdrop } from './Backdrop';
