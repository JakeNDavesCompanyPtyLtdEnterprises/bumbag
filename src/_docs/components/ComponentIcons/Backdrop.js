import React from 'react';

const SvgComponent = props => (
  <svg width={120} height={60} fill="none" {...props}>
    <path fill="#4F4F4F" d="M0 0h120v60H0z" />
  </svg>
);

export default SvgComponent;
