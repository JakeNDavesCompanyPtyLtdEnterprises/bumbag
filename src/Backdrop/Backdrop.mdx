---
name: Backdrop
route: /utils/backdrop
menu: Utilities
---

import { Playground, PropsTable } from 'docz';
import { Box } from '../primitives/index';
import Backdrop from './index';
import Button from '../Button';
import Pane from '../Pane';
import Overlay from '../Overlay';

# Backdrop

## Import

`import { Backdrop } from 'fannypack'`

## Basic Usage

`<Backdrop>` uses the [`<Hidden>` component](/utils/hidden) internally, and adds click functionality (as it is a button).

In the example below, we are leveraging the [`<Overlay>` component](/utils/overlay) to help us to show/hide the backdrop.

<Playground>
  <Overlay.Container>
    {({ visible, show, hide, toggle }) => (
      <Box>
        <Button onClick={show}>Show backdrop</Button>
        <Backdrop visible={visible} />
        <Overlay visible={visible}>
          <Button onClick={hide}>Hide backdrop</Button>
        </Overlay>
      </Box>
    )}
  </Overlay.Container>
</Playground>

## Animations

You can also add animations to backdrop if you like.

### Expand

Have an expand-like animation by using the `expand` prop. The `expand` prop can take a boolean, or a position of `center`, `left`, `right`, `top` or `bottom`.

<Playground>
  <Overlay.Container>
    {({ visible, show, hide, toggle }) => (
      <Box>
        <Button onClick={show}>Show backdrop</Button>
        <Backdrop expand visible={visible} />
        <Overlay visible={visible}>
          <Button onClick={hide}>Hide backdrop</Button>
        </Overlay>
      </Box>
    )}
  </Overlay.Container>
</Playground>

### Fade

Have a fade animation by using the `fade` prop.

<Playground>
  <Overlay.Container>
    {({ visible, show, hide, toggle }) => (
      <Box>
        <Button onClick={show}>Show backdrop</Button>
        <Backdrop fade visible={visible} />
        <Overlay visible={visible}>
          <Button onClick={hide}>Hide backdrop</Button>
        </Overlay>
      </Box>
    )}
  </Overlay.Container>
</Playground>

### Slide

You can have a slide animation with the `slide` prop. The `slide` prop can take a boolean, or a position of `left`, `right`, `top` or `bottom`.

<Playground>
  <Overlay.Container>
    {({ visible, show, hide, toggle }) => (
      <Box>
        <Button onClick={show}>Show backdrop</Button>
        <Backdrop slide="bottom" visible={visible} />
        <Overlay visible={visible}>
          <Button onClick={hide}>Hide backdrop</Button>
        </Overlay>
      </Box>
    )}
  </Overlay.Container>
</Playground>

## Props

<PropsTable of={Backdrop} />

## Theming

### Schema

```jsx
{
  base: string | Object;
}
```
