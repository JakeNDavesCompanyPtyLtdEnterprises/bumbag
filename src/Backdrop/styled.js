// @flow
import styled from 'reakit/styled';
import { theme } from 'styled-tools';
import Backdrop from 'reakit/Backdrop';

export default styled(Backdrop)`
  background-color: black;
  opacity: 0.4;

  & {
    ${theme('fannypack.Backdrop.base')};
  }
`;
