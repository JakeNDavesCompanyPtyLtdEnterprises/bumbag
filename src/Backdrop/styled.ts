// @ts-ignore
import styled from 'reakit/styled';
import { theme } from 'styled-tools';
// @ts-ignore
import Backdrop from 'reakit/Backdrop';

export default styled(Backdrop)`
  background-color: black;
  opacity: 0.4;

  & {
    ${theme('fannypack.Backdrop.base')};
  }
`;
