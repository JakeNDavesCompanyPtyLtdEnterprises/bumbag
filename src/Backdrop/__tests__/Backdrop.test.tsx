import * as React from 'react';
// @ts-ignore
import render from '../../_utils/tests/render';
import Backdrop from '../Backdrop';

it('renders correctly for a basic backdrop', () => {
  const { container } = render(<Backdrop isVisible>test</Backdrop>);
  expect(container.firstChild).toMatchSnapshot();
});
