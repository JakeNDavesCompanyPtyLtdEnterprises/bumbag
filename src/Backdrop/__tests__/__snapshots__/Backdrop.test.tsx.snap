// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for a basic backdrop 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  -webkit-transform: translate3d(0px,0px,0px);
  -ms-transform: translate3d(0px,0px,0px);
  transform: translate3d(0px,0px,0px);
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 998;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  -moz-tap-highlight-color: rgba(0,0,0,0);
  background-color: black;
  opacity: 0.4;
}

.c0[aria-hidden="true"] {
  pointer-events: none;
  display: none !important;
}

<div
  aria-hidden="false"
  class="c0 c1"
  role="button"
  tabindex="-1"
>
  test
</div>
`;
